examples/i3-cycle-focus.py

```diff
@@ -5,100 +5,168 @@
         self.i3 = await Connection().connect()
         self.i3.on('window::focus', self.on_window_focus)
-        self.i3.on('shutdown', on_shutdown)
+        self.i3.on('shutdown', self.on_shutdown)
 
-    async def update_window_list(self, window_id):
+    async def update_window_list(self, container):
         if UPDATE_DELAY != 0.0:
             await asyncio.sleep(UPDATE_DELAY)
 
         logging.info('updating window list')
-        if window_id in self.window_list:
-            self.window_list.remove(window_id)
 
-        self.window_list.insert(0, window_id)
+        key = GLOBAL_KEY
+        if KEYED_CONF:
+            key = (container.ipc_data['output'] if PER_OUTPUT
+                   else str((await self.i3.get_tree()).find_focused().workspace().num))
+
+        wlist = self.window_list.get(key)
+        if wlist is None:
+            wlist = self.window_list[key] = []
+            self.window_index[key] = [1]
+        else:
+            self.window_index[key][0] = 1
+
+        window_id = container.window
+        if window_id in wlist:
+            wlist.remove(window_id)
 
-        if len(self.window_list) > MAX_WIN_HISTORY:
-            del self.window_list[MAX_WIN_HISTORY:]
+        wlist.insert(0, window_id)
 
-        self.window_index = 1
-        logging.info('new window list: {}'.format(self.window_list))
+        if len(wlist) > MAX_WIN_HISTORY:
+            del wlist[MAX_WIN_HISTORY:]
 
-    async def get_valid_windows(self):
-        tree = await self.i3.get_tree()
-        if args.active_workspace:
-            return set(w.id for w in tree.find_focused().workspace().leaves())
+        logging.info('new window list: {}'.format(wlist))
+
+    async def get_valid_windows(self, tree, focused_ws):
+        if args.active_workspace or args.focused_workspace:
+            return set(w.window for w in focused_ws.leaves())
         elif args.visible_workspaces:
             ws_list = []
             w_set = set()
             outputs = await self.i3.get_outputs()
-            for item in outputs:
-                if item.active:
-                    ws_list.append(item.current_workspace)
+            for output in outputs:
+                if output.active:
+                    ws_list.append(output.current_workspace)
             for ws in tree.workspaces():
                 if str(ws.name) in ws_list:
                     for w in ws.leaves():
-                        w_set.add(w.id)
+                        w_set.add(w.window)
+            return w_set
+        elif args.focused_output:
+            w_set = set()
+            focused_output = focused_ws.ipc_data['output']
+            for ws in tree.workspaces():
+                if ws.ipc_data['output'] == focused_output:
+                    for w in ws.leaves():
+                        w_set.add(w.window)
             return w_set
         else:
-            return set(w.id for w in tree.leaves())
+            return set(w.window for w in tree.leaves())
```

Comment:
Note we track `window` property instead of `id`, as it stays constant across i3 restarts.